An Infrastructure for Extendable Input/Output File Format Support in DAPHNE

yazanandoni opened the pull request.

yazanandoni committed:
27.05

yazanandoni committed:
added the option to put costume options as a frame

yazanandoni committed:
before changing read

yazanandoni committed:
added extendability arguement in the comand line

yazanandoni committed:
use of one registry by utelizing the daphne context

yazanandoni committed:
added all built-in readers as a plug in

yazanandoni committed:
added benchmarks and created a clear function for the Registry

yazanandoni committed:
changes and fixes

yazanandoni committed:
added writer support and changed write.h

yazanandoni committed:
added library loading when plug-in is used and changed the write kernel

yazanandoni committed:
seperated the benchmarks and changed the csv plugin

yazanandoni committed:
fixed some bugs

yazanandoni committed:
engine and priority support

yazanandoni committed:
removed built-ins and made them only as a plug-in that gets loaded automatically when daphne starts

yazanandoni committed:
added the script tests

yazanandoni committed:
adjustments

yazanandoni committed:
added eval_runner and cleaned some code

yazanandoni committed:
final commit

yazanandoni committed:
final commit

pdamme reviewed (changes requested):
**Thank you so much for this PR, @yazanandoni**, and sorry for the delay. The **contribution of this PR** is twofold:

1. An extensibility infrastructure for file formats (readers/writers as plug-ins)
1. Concrete efficient plug-ins for CSV and Parquet

The **extensibility infrastructure** is **highly valuable to DAPHNE** as it can help us (and expert users) to easily add support for more file formats without requiring more and more dependencies for the main system. The **concrete plug-ins** are **also highly welcome** as they can improve the performance of reads/writes in DAPHNE.

-----

**The following comments refer only to the extensibility infrastructure**, which I would like to merge first as it enables follow-up work on more plug-ins. *The concrete plug-ins will be handled later.*

The code of the proposed extensibility infrastructure is **largely fine, but several improvements and corrections are required** before we can merge it in.

**Mandatory points:**
1. All files must adhere to the code style to make the CI checks pass.
1. There are some smaller issues with the extensibility infrastructure, which should be fixed, e.g.:
    - Don't read the plug-in priority from the JSON file, since the priority is inherently related to the registration of the plug-in, not to the plug-in itself.
    - `FileIORegistry::getReader()`/`getWriter()` may not take the priority into account correctly. They would prefer a low-priority reader/writer (plug-in that has already been loaded) over high-priority lazy specification (plug-in that has not been loaded yet).
    - `registerLazy()` should ensure that at least one of `readerSymbol`/`writerSymbol` is specified.
    - `Read.h`/`Write.h`: HDFS is handled after the plug-ins and will most likely not work anymore.
    - Do we really need a mutex for the access to the catalog?
    - Readers and writers use the same options map, and could overwrite each other.
1. The test cases need some improvements, e.g.:
    - The tests cases for the extensibility infrastructure should not depend on concrete plug-ins and data files in `scripts/examples/`, they should only use test files from the `test/` directory.
    - The test output files should be `.gitignore`'d.
    - The new test cases should focus on the extensibility infrastructure, not on concrete new extensions.
    - The test cases should not use Arrow (or other third-party libraries) in order to avoid additional dependencies.
    - Some test cases print debug messages, these must be silenced.
1. The code needs a thorough polishing pass, e.g.:
    - Polishing identifiers, comments, and error messages to make them more understandable and more consistent with the rest of the code base.
    - Ensure correct license headers in all files.
    - Undo several unrelated changes.
1. Refactoring of the directory and file structure of the new plug-ins (both for built-in readers/writers and new custom ones), e.g.:
    - Currently, these plug-ins reside in `scripts/examples/`. However, they are not just usage examples, but actual reusable code. Maybe a new directory `extensions/` would be more appropriate for them.
    - Currently, the built-in plug-ins consists of a large cpp file, which seems to concatenate the code of the original built-in readers/writers from `src/runtime/local/io/`. Retaining the separation into multiple files would be a clearer structure.
    - The `FileIOCatalogParser` should be moved from to `src/parser/` (where all other parsers reside, including those for the kernel extension catalog).
1. Several artifacts that we don't need on the main branch need to be removed, e.g., experiments and compiled binaries.

**Performance impact.** I conducted some little experiments on file read/write performance (writing randomly generated dense and sparse matrices to CSV and the DAPHNE Binary Data Format, writing randomly generated frames with and without string columns to CSV, reading these matrices and frames again, and reading sparse matrices from MatrixMarket). Consistently with what you reported in you thesis, I observed that reading via a plug-in yields only a negligible overhead compared to the built-in readers, but writing is noticeably slower when done via a plug-in (slow-down of roughly 10-20% for numerical data and 30-100% for string data). *We should clearly understand (and ideally fix) these performance issues before we make plug-ins the default path for reading/writing files.*

-----

**I will finalize the code in this PR and merge it in (initially only the extensibility infrastructure).**

pdamme committed:
fix: parsing DaphneDSL read/write built-ins with optional extra options.

- This fix became necessary after rebasing this branch on the upstream main branch.

pdamme committed:
style: applied clang-format to all h/cpp files touched by this PR

pdamme committed:
fix: removed unwanted debug output

pdamme committed:
style: make CI code style checks pass

- There seem to be some inconsistencies between my local clang-format and the one in the CI container.

pdamme committed:
perf: harmonize compiler flags used for built-in write-kernel and built-in IO extensions

- Initial experiments showed that using the source code of DAPHNE's CSV readers/writers through IO plug-ins can yield a significant runtime overhead (especially for writing string-heavy data).
- A deeper investigation revealed that this overhead can be attributed to a subtle difference in the compiler flags used when compiling the built-in write-kernel and the IO extension.
- More precisely, the write-kernel uses `-std=gnu++20`, while the extension used `-std=c++17`, which can apparently cause a significant performance difference.
- This commit harmonizes the compiler flags by also using `-std=gnu++20` for the extension.
- With that, the runtime overhead mentioned above vanishes.

pdamme committed:
style: clean-up pass over src directory

- General things:
  - Added missing license headers.
  - Grouped and sorted #includes.
  - Renamed misleading identifiers (e.g., prompPriority to priority, typeHash to dt (for data type), IODataType to PhyDataType).
  - Turned IOOptions from a struct into a using definition, since it is basically just an alias for a map from string to string.
  - Adjusted various error messages.
  - Removed various unnecessary comments and commented-out code lines.
- daphne_internal.cpp:
  - Harmonized parsing of file IO extensions and kernel extensions.
  - Renamed CLI arg for file IO extension from --FileIO-ext to --fileio-ext for consistency (lower case).
- DaphneDSL parser:
  - Simplified parsing of readMatrix()/readFrame()/write() built-ins: Instead of creating a 1x1 dummy frame, create an empty 1x0 frame (made possible through a new commit on main).
- FileIOCatalogParser::parseFileIOCatalog():
  - Parameter `priority` is not optional anymore (callers may pass 0 if they want).
  - Harmonized the implementation with `KernelCatalogParser::parseKernelCatalog()`.
- FileIORegistry:
  - Removed the old "backward-compatible" registration functions.
  - Harmonized the parameter order of registerReader()/registerWriter()/registerLazy().
  - Changed std::out_of_range to std::runtime_error, since we virtually always use std::runtime_error.
- Read/write-ops/kernels:
  - Changed type of options argument of DaphneIR ReadOp and WriteOp from MatrixOrFrame to FrameOrU, since it is always a frame.
  - Made the kernels' options parameter a const frame (we don't need to change the options).
- And several other minor clean-ups.

pdamme committed:
refactor: Write.h does not include Read.h anymore

- Write.h used to include Read.h to get access to the two helper functions mergeOptionsFromFrame() and extractEngineFromFrame(); however, it does not need the actual read-kernel.
- Moved these two helper functions to FileIORegistry.h, which was split into h/cpp files.
- Now both Read.h and Write.h get the helper functions from FileIORegistry.h.

pdamme committed:
fix: empty options frames are accepted again

pdamme committed:
refactor: renamed FileIORegistry to FileIOCatalog

- For consistency with KernelCatalog and with the existing FileIOCatalogParser.
- Renamed FileIORegistry.h/cpp to FileIOCatalog.h/cpp.
- Renamed identifiers throught the code base (e.g., registry to catalog, reg to cat, etc.).
- And some other little related changes.

pdamme committed:
fix: disallow "engine" for extension-specific option names

- The key "engine" can be used to select a particular file IO extension, e.g., with DaphneDSL's readMatrix()/readFrame()/write().
- As this key is passed to the kernels in the same options frame as extension-specific options, this name should not be allowed for extension-specific options to avoid name clashes.
- This commit ensures during file IO catalog parsing that no extension-specific option is called "engine".
- Furthermore, when merging default options and user-specified options, a potentially speficied "engine" key is ignored (already used to be that way).
- Hence, a particular extension will not see the "engine" key.
- At the same time, we don't need to exclude the key "priority" in any way, because the priority of a file IO extension is only specified at catalog parse-time, but not when passing extension-specific options around (thus, an extension-specific option called "priority" wouldn't hurt, even though it might be slightly missleading).

pdamme committed:
refactor: remove FileIOCatalog::getReader()/getWriter() overloads

- These two methods had overloads that didn't require specifying the engine.
- While those can be convenient, they are not necessary (one can pass "" for the engine).

pdamme committed:
fix: exception handling in read/write-kernels

- A large part of the read/write-kernels' function bodies was covered by a try-catch that threw on an exception reporting that no suitable reader/writer was found in case of an exception.
- However, an exception inside the kernel could have various reasons, e.g., it might be the case that a suitable reader/writer was indeed found but its shared lib could not be opened.
- This commit improves the exception handling in the read/write-kernels as follows:
  - The try-block covers the entire kernel function body (except for the HDFS-related part in the DenseMatrix-kernels, that will be fixed soon), to handle all potential sources of errors in a unified way.
  - The catch-block catches the superclass std::exception, to handle all kinds of errors in a unified way.
  - In case of an error, a new std::runtime_error is thrown, which (1) reports that the error happened while reading/writing the specified file (will be helpful context for users), and (2) the message of the caught exception.

pdamme committed:
fix: treatment of HDFS in read/write-kernels

- So far, reading/writing HDFS files has not been ported to file IO extensions.
- Instead, the read/write-kernels invoke readHDFS()/writeHDFS() directly.
- This commit makes sure that HDFS is correctly covered by the new try-catch statement.
- Eventually, we should also support HDFS through a file IO extension.

pdamme committed:
fix: physical data type assumed by write-kernel on abstract Matrix

- The specialization of the write-kernel for the abstract Matrix superclass assumed the physical data type PhyDataType::CSRMATRIX for selecting a suitable writer from the catalog.
- However, this assumption is not generally true (we don't statically know what implementation is behind the abstract Matrix).
- This commit adds a check for the actual physical data type (using dynamic_cast).

pdamme committed:
fix: accessing options of various value types in mergeOptionsFromFrame()

- Extension-specific options can be of various value types (e.g., the separator in CSV is a string, while the number of threads to use is an integer).
- mergeOptionsFromFrame() converts all user-specified options to strings for simpler internal handling.
- However, there was a bug when accessing frame columns of various types: The function tried to get the column as a specific value type and then check via a dynamic_cast if it really is that type.
- That doesn't work because already when getting the column, the frame internally checks if the value type matches the requested type and throws, otherwise.
- Furthermore, the dynamic_cast was superfluous, since it casted to exactly the type that Frame::getColumn<VT>() anyway returns.
- The only reason why it worked so far is that the value type std::string was checked first and all examples in the test cases where strings (one could easily provoke test failures by changing the order of the checks).
- This commit fixes this bug by first checking the column's value type and then extracting the column as the right value type.

pdamme committed:
style: simplified identifiers related to options

- `optsFrame` to `opts`, `mergeOptionsFromFrame()` to `mergeOptions()`, `extractEngineFromFrame()` to `extractEngine()`.
- Simpler, still expressive enough, and not focused on the fact that we use a frame to represent the options.
